### PR TITLE
Add macro to simplify libgit2 type constructor definitions

### PR DESCRIPTION
**Pull request by simonbyrne**
This adds a macro that automatically defines a keyword argument only constructor for a type declaration.

### CODE REVIEW

**Comment by simonbyrne** [comment]
I also changed a couple of fields from `Cint` to specific enums.

File: base/libgit2/types.jl [hunk 1]
```diff
@@ -1,6 +1,42 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-const Cstring_NULL = convert(Cstring, Ptr{UInt8}(C_NULL))
+"""
+    @kwdef typedef
+
+This is a helper macro that automatically defines a keyword-based constructor for the type declared in the expression `typedef`, which must be a `type` or `immutable` expression. The default argument is supplied by declaring fields of the form `field::Type = default`. If no default is provided then the default is:
```
Comment (ararslan) [inline base/libgit2/types.jl]: Perhaps this should be broken onto multiple lines?

**Comment by tkelman** [comment]
This should probably be in util.jl or something since it's not really libgit2 specific.

**Comment by StefanKarpinski** [comment]
I like this idea. We should probably have language features for this kind of thing eventually.

**Comment by ararslan** [comment]
If we're going to have it be more general and user-facing, we might want to make the name a little more specific. I don't have a better suggestion though...

**Comment by simonbyrne** [comment]
I've moved it, open to better suggestions on names.

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -487,80 +299,67 @@ immutable IndexEntry
...
 Base.show(io::IO, ie::IndexEntry) = print(io, "IndexEntry($(string(ie.id)))")
 
+"""
+    LibGit2.RebaseOptions
 
-if LibGit2.version() >= v"0.24.0"
-    immutable RebaseOptions
-        version::Cuint
-        quiet::Cint
+Matches the `git_rebase_options` struct.
```
Comment (simonbyrne) [inline base/libgit2/types.jl]: No, their HTML docs are a bit incomplete.

**Comment by simonbyrne** [comment]
Any more thoughts on this?

**Pull request merged**